Initial config for remote debugging (#5030)
nginxinc/kubernetes-ingress: docs/developer/debugging.md

    controller:
        debug:
            enable: true
        kind: daemonset
        service:
            type: NodePort
            customPorts:
              - name: godebug
                nodePort: 32345
                port: 2345
                protocol: TCP
                targetPort: 2345
        customPorts:
          - name: godebug
            containerPort: 2345
            protocol: TCP
        readyStatus:
            enable: false
        image:
            tag: debug
            repository: local/nic-debian
    EOF
    ```
6. Add the following launch configuration to the `configurations` section of your VSCode `.vscode/launch.json` or equivalent for your IDE of choice:
    ```json
    {
        "name": "Debug NIC in local Kind cluster",
        "type": "go",
        "request": "attach",
        "mode": "remote",
        "remotePath": "",
        "port":32345,
        "host":"localhost",
        "showLog": true,
        "cwd": "${workspaceFolder}"
    }
    ```
7. Run the configuration from the `Run and Debug` menu, set some breakpoints, and start debugging!

## Debug configuration walkthrough

### 1. Build a debug container image

> Build a NIC container with either:

Reviewer: ```suggestion
Create an NGINX Ingress Controller container with either:
```

1. `make <image name> TARGET=debug`

> This builds the debuggable NIC binary locally and loads it into the container image

Reviewer: ```suggestion
This option creates the debug NGINX Ingress Controller binary locally , then loads it into the container image.
```

1. `make <image name> TARGET=debug-container`

> This builds the debuggable NIC binary in the container image

Reviewer: ```suggestion
This options builds the debug NGINX Ingress Controller binary directly inside the container image.
```

> The debug image will use a NIC binary which contains debug symbols and has [Delve](https://github.com/go-delve/delve) installed. This image also uses `/dlv` as the entrypoint.

Reviewer: ```suggestion
The debug NGINX Ingress Controller binary contains debug symbols and has [Delve](https://github.com/go-delve/delve) installed. It uses `/dlv` as its entrypoint.
```

> Example for building a Debian image with NGINX Plus on Arm64 which will be tagged as `local/nic-debian-plus:debug`:

Reviewer: ```suggestion
The following example builds a Debian image with NGINX Plus on ARM64,  tagged as `local/nic-debian-plus:debug`:
```

```shell
make debian-image-plus TARGET=debug PREFIX=local/nic-debian-plus TAG=debug ARCH=arm64
...
...
 => => naming to docker.io/local/nic-debian-plus:debug
```

### 2. Deploy the debug container

> Enable the debug configuration via Helm:

Reviewer: ```suggestion
Use Helm to enable the debug configuration:
```

```yaml
controller:
    debug:
        enable: true
    service:
        type: NodePort
        customPorts:
        # only required if you want to connect
        # directly to your cluster instead of using kubectl port-forward
          - name: godebug
            nodePort: 32345
            port: 2345
            protocol: TCP
            targetPort: 2345
    customPorts:
      - name: godebug
        containerPort: 2345
        protocol: TCP
    readyStatus:

>     # it is recommended to deactivate readinessProbes while debugging

Reviewer: ```suggestion
    # We recommend deactivating readinessProbes while debugging
```

    # to ensure upgrades and service connections run as expected
        enable: false
    image:
        tag: debug
        repository: local/nic-debian-plus
```

> Or if not using Helm manually add the Delve CLI flags to the deployment or daemonset:

Reviewer: ```suggestion
If you are not using Helm, manually add the Delve CLI arguments to the deployment or daemonset:
```

```yaml
args:
- --listen=:2345
- --headless=true
- --log=true
- --log-output=debugger,debuglineerr,gdbwire,lldbout,rpc,dap,fncall,minidump,stack
- --accept-multiclient
- --api-version=2
- exec
- ./nginx-ingress
- --continue
- --

> <regular NIC CLI configuration>

Reviewer: ```suggestion
<regular NGINX Ingress Controller CLI configuration>
```

```

> By default Delve will immediately start NIC. Setting `controller.debug.continue: false` will cause Delve to wait for a debugger to connect before starting NIC. This is useful for debugging startup behavior of NIC.

Reviewer: ```suggestion
Delve will immediately start NGINX Ingress Controller by default. To prevent this, set `controller.debug.continue: false.`  Delve will then wait for a debugger to connect before starting NGINX Ingress Controller, which is useful for observing start-up behaviour.
```

### 3. Connect your debugger

Connect to the remote Delve API server through your IDE:
- [JetBrains](https://www.jetbrains.com/help/go/attach-to-running-go-processes-with-debugger.html)
- [VSCode](https://github.com/golang/vscode-go/blob/master/docs/debugging.md)

Example VSCode configuration:

```json
{
    "name": "Debug NIC",
    "type": "go",
    "request": "attach",
    "mode": "remote",
    "remotePath": "",
    "port":32345,
    "host":"<cluster where nodeport is exposed, or localhost if using kubectl port forward>",
    "showLog": true,
    "cwd": "${workspaceFolder}"
}
```

> You may want to expose the debugging port on your cluster via `kubectl port-forward`, for example using:

Reviewer: ```suggestion
You may want to expose the debug port of your cluster via `kubectl port-forward` :
```
